trieste/models/optimizer.py

```diff
@@ -139,65 +152,51 @@ def train_fn() -> None:
 
 @singledispatch
 def create_optimizer(
-    optimizer: gpflow.optimizers.Scipy | tf.optimizers.Optimizer,
+    optimizer: Any,
     optimizer_args: Dict[str, Any],
 ) -> Optimizer:
     """
     Generic function for creating a :class:`Optimizer` wrapper from a specified
     `optimizer` and `optimizer_args`. The implementations depends on the type of the
-    underlying optimizer.
+    underlying optimizer, which can be model-specific. This function needs to be used as a
+    decorator together with its register method to make a (model-)specific optimizer available.
 
-    :param optimizer: The optimizer with which to train the model.
-    :param optimizer_args: The keyword arguments to pass to the optimizer wrapper..
+    :param optimizer: The optimizer with which to train the model. Note that a flexible type
+        `Any` is used to allow for vartious optimizers that specific models might need to use.
+    :param optimizer_args: The keyword arguments to pass to the optimizer wrapper.
     :return: The :class:`Optimizer` wrapper.
     """
+    raise NotImplementedError(f"Unknown optimizer {optimizer} passed for creating an optimizer")
```

Comment:
That's a good catch about not implemented. But not sure about `Any`. Why does the type need to change at all?

Reply:
see my comment above

Reply:
`NotImplementedError` wasn't required before since there were overloads for all the supported types (SciPy and Optimizer). Changing it to `Any` means you now need it.